// d360/dist/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandCategories = exports.ciNames = exports.AllCommandDetails = exports.Messages = exports.DefaultMainArgs = void 0;
exports.DefaultMainArgs = [
    { name: 'help', alias: 'h', type: Boolean },
    { name: 'command', type: String, defaultOption: true }
];
exports.Messages = {
    apiKeyNotFound: "🗝️ API Key not found",
    apiKeyNotFoundInCI: "🗝️ API Key not found. Please use --apiKey",
    userIdNotFound: "👨‍💻 User Id not found",
    userIdNotFoundInCI: "👨‍💻 User Id not found. Please use --userId"
};
exports.AllCommandDetails = [
    {
        commandCategoryName: "apidocs",
        decsription: "API Documentation",
        commands: []
    }
];
exports.ciNames = {
    'aws-codebuild': 'AWS CodeBuild',
    'azure-pipelines': 'Azure Pipelines',
    'bamboo': 'Bamboo',
    'bitbucket-pipelines': 'Bitbucket Pipelines',
    'bitrise': 'Bitrise',
    'buddy': 'Google Cloud Builder',
    'buildkite': 'Buildkite',
    'circleci': 'Circle-CI',
    'cirrus': 'Cirrus CI',
    'codeship': 'CodeShip',
    'custom': 'Custom',
    'drone': 'Drone',
    'dsari': 'dsari CI',
    'gerrit': 'gerrit',
    'github-actions': 'GitHub Actions',
    'gitlab': 'GitLab',
    'gocd': 'GoCD',
    'heroku': 'Heroku',
    'hudson': 'Hudson CI',
    'jenkins': 'jenkins',
    'magnum': 'Magnum CI',
    'netlify': 'Netlify',
    'nevercode': 'Nevercode',
    'now': "Zeit.co's Now service, but not GitHub/BitBucket/GitLab",
    'now-bitbucket': "Zeit.co's Now for BitBucket deployment service",
    'now-github': "Zeit.co's Now for GitHub deployment service",
    'now-gitlab': "Zeit.co's Now for GitLab deployment service",
    'render': 'Render CI',
    'sail': 'Sail CI',
    'screwdriver': 'Screwdriver CI',
    'semaphore': 'Semaphore',
    'shippable': 'Shippable',
    'strider': 'Strider CI',
    'taskcluster': 'Mozilla Taskcluster',
    'tddium': 'TDDium',
    'teamcity': 'TeamCity',
    'travis-ci': 'Travis-CI',
    'vercel': 'Vercel',
    'vercel-bitbucket': 'Vercel Bitbucket',
    'vercel-github': 'Vercel GitHub',
    'vercel-gitlab': 'Vercel Gitlab',
    'wercker': 'Oracle Wercker',
    'woodpecker': 'Woodpecker CI'
};
var CommandCategories;
(function (CommandCategories) {
    CommandCategories["apidocs"] = "apidocs";
})(CommandCategories = exports.CommandCategories || (exports.CommandCategories = {}));
