// d360/dist/src/commands/baseCommand.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var setConfig_1 = require("../api/setConfig");
var constants_1 = require("../constants");
var consoleWrapper_1 = require("../helper/consoleWrapper");
var isCI_1 = require("../helper/isCI");
var terminalWrapper_1 = __importDefault(require("../helper/terminalWrapper"));
var BaseCommand = /** @class */ (function () {
    function BaseCommand() {
    }
    /*
      The arguments that are required for all commands and that doesn't have a default value,
      we should validate it in the base command
    */
    BaseCommand.prototype.run = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!options.apiKey) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.getApiKey(options)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!!options.userId) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.getUserId(options)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/, Promise.resolve("")];
                }
            });
        });
    };
    BaseCommand.prototype.getApiKey = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var apiKeyAnswer;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!(0, isCI_1.isCI)()) return [3 /*break*/, 2];
                        (0, consoleWrapper_1.error)(constants_1.Messages.apiKeyNotFound);
                        return [4 /*yield*/, (0, terminalWrapper_1.default)([
                                {
                                    type: 'text',
                                    name: 'apiKey',
                                    message: 'Enter your API Key',
                                }
                            ])];
                    case 1:
                        apiKeyAnswer = _a.sent();
                        if (apiKeyAnswer.apiKey) {
                            options.apiKey = apiKeyAnswer.apiKey;
                            (0, setConfig_1.setAPIKey)(apiKeyAnswer.apiKey);
                        }
                        return [3 /*break*/, 3];
                    case 2: throw new Error(constants_1.Messages.apiKeyNotFoundInCI);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    BaseCommand.prototype.getUserId = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var apiKeyAnswer;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        console.log(options.userId);
                        if (!!(0, isCI_1.isCI)()) return [3 /*break*/, 2];
                        (0, consoleWrapper_1.error)(constants_1.Messages.userIdNotFound);
                        return [4 /*yield*/, (0, terminalWrapper_1.default)([
                                {
                                    type: 'text',
                                    name: 'userId',
                                    message: 'Enter user Id',
                                }
                            ])];
                    case 1:
                        apiKeyAnswer = _a.sent();
                        if (apiKeyAnswer.userId) {
                            options.userId = apiKeyAnswer.userId;
                            (0, setConfig_1.setUserId)(apiKeyAnswer.userId);
                        }
                        return [3 /*break*/, 3];
                    case 2: throw new Error(constants_1.Messages.apiKeyNotFoundInCI);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return BaseCommand;
}());
exports.default = BaseCommand;
