// d360/dist/src/api/apidocs/validate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
//import chalk from "chalk";
var form_data_1 = __importDefault(require("form-data"));
var fs_1 = __importDefault(require("fs"));
var node_fetch_1 = require("node-fetch");
var ora_1 = __importDefault(require("ora"));
var consoleWrapper_1 = require("../../helper/consoleWrapper");
var isURL_1 = require("../../helper/isURL");
var d360APIFetch_1 = __importDefault(require("../d360APIFetch"));
function validateFlow(options) {
    return __awaiter(this, void 0, void 0, function () {
        var spinner, formData, requestOptions, requestUrl, response;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    spinner = (0, ora_1.default)({ text: "Validating.." });
                    spinner.start();
                    formData = getImportRequestFormData(options);
                    requestOptions = {
                        method: "POST",
                        body: formData,
                        headers: new node_fetch_1.Headers({
                            api_token: options.apiKey,
                        }),
                    };
                    requestUrl = "/v2/apidocs/validate";
                    return [4 /*yield*/, (0, d360APIFetch_1.default)(requestUrl, requestOptions)];
                case 1:
                    response = _a.sent();
                    spinner.stop();
                    handleValidateResponse(response);
                    return [2 /*return*/];
            }
        });
    });
}
exports.default = validateFlow;
function getImportRequestFormData(options) {
    var formData = new form_data_1.default();
    if (!(0, isURL_1.isURL)(options.path)) {
        var stream = fs_1.default.createReadStream(options.path);
        formData.append("file", stream);
    }
    else {
        formData.append("url", options.path);
    }
    return formData;
}
function showAlerts(importWarnings) {
    (0, consoleWrapper_1.heading)("\n");
    (0, consoleWrapper_1.heading)("Here are the alert(s) found in the OpenAPI spec file.");
    for (var i = 0; i < importWarnings.length; i++) {
        (0, consoleWrapper_1.warning)("Message :" + importWarnings[i]["message"] + "\n" + "Path :" + importWarnings[i]["pointer"] + "\n");
    }
}
function showErrors(importErrors) {
    (0, consoleWrapper_1.heading)("Here are the errors found in the OpenAPI spec file.");
    for (var i = 0; i < importErrors.length; i++) {
        (0, consoleWrapper_1.error)("Message :" + importErrors[i]["message"] + "\n" + "Path :" + importErrors[i]["pointer"] + "\n");
    }
}
function handleValidateResponse(response) {
    var _a, _b;
    if (response === null || response === void 0 ? void 0 : response.success) {
        var importWarnings = (_a = response === null || response === void 0 ? void 0 : response.result) === null || _a === void 0 ? void 0 : _a.warnings;
        var importErrors = (_b = response === null || response === void 0 ? void 0 : response.result) === null || _b === void 0 ? void 0 : _b.errors;
        if ((importErrors == null && importWarnings == null) || (importErrors.length <= 0 && importWarnings.length <= 0)) {
            (0, consoleWrapper_1.success)("No errors were found. Your spec file has been validated successfully.");
        }
        else if (importWarnings.length > 0 && importErrors.length > 0) {
            showAlerts(importWarnings);
            showErrors(importErrors);
            (0, consoleWrapper_1.error)("Validation failed! We found ".concat(importWarnings.length, " alert(s) and ").concat(importErrors.length, " errors."));
        }
        else if (importWarnings.length > 0) {
            showAlerts(importWarnings);
            (0, consoleWrapper_1.warning)("Validation failed! We found ".concat(importWarnings.length, " alert(s)."));
        }
        else if (importErrors.length > 0) {
            showErrors(importErrors);
            (0, consoleWrapper_1.error)("Validation failed! We found ".concat(importErrors.length, " errors."));
        }
    }
}
