// d360/dist/src/api/getConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUserId = exports.getAPIKey = exports.getApiHubUrl = void 0;
var config_store_1 = __importDefault(require("./config-store"));
/**
 * Retrieves stored user data values from env variables or configstore,
 * with env variables taking precedent
 */
function getConfig() {
    var environmentConfig = {
        apiKey: process.env.D360_API_KEY || config_store_1.default.get('apiKey'),
        apihubUrl: process.env.D360_APIHUB_URL || config_store_1.default.get('apihubUrl')
    };
    return environmentConfig;
}
exports.default = getConfig;
function getApiHubUrl() {
    return process.env.D360_APIHUB_URL || config_store_1.default.get('apihubUrl') || config_store_1.default.get("env.eu_baseUrl");
}
exports.getApiHubUrl = getApiHubUrl;
function getAPIKey() {
    return process.env.D360_API_KEY || config_store_1.default.get("apiKey");
}
exports.getAPIKey = getAPIKey;
function getUserId() {
    return process.env.D360_USER_ID || config_store_1.default.get("userId");
}
exports.getUserId = getUserId;
